added openrouter api and loaded the api key if exist in .env file

rsrini7 opened the pull request:
1. **OpenRouter Provider Integration**
   - Added OpenRouter as a supported LLM provider.
   - Integrated OpenRouter with OpenAI-compatible API using the LangChain SDK.
   - Initial implementation used a simple HTTP layer, now upgraded to use LangChain for better maintainability and features.

2. **Automatic API Key Management**
   - Implemented logic to auto-load the API key from `.env` for the selected LLM provider.
   - The API key input field is now auto-filled and disabled if the key is present in `.env`, streamlining the user experience.
   - Session state and Streamlit widget handling improved to avoid warnings and ensure seamless switching between providers.

3. **Streamlit UI Enhancements**
   - Removed Streamlit widget warnings related to conflicting session state and widget values.
   - Ensured robust and user-friendly session state management for API keys.

4. **Configuration Example**
   - Added a default [.env.example](cci:7://file:///Users/srini/Ws/slide-deck-ai/.env.example:0:0-0:0) file to the repository.
   - Provides a template for required environment variables, including `OPENROUTER_API_KEY`, `OPENROUTER_SITE_URL`, and `OPENROUTER_SITE_NAME`.
   - Helps users quickly configure their environment for OpenRouter and other providers.
  
5. Dynamically added input text limit in streamlit ui and removed .idea folder

barun-saha commented:
Hi Srinivasan,

Thanks for your interest and contributions!

I'll have a detailed look later and try running the code. Meanwhile, a few observations after a glance:

- How does Mixtral 8x22B Instruct perform? Does it generate useful content for the slides? Does it always follow the output JSON schema? I haven't tried it, so I want to know.
- OpenRouter is also new to me. In particular, I think I noticed `localhost` somewhere in the changes, so need to check how it works.
- Mapping API keys from the `.env` file is nice. It can be helpful for testing. I'll take a look later.
- Finally, the changes seem to introduce an input field to specify the input text size. I'm not sure it is required. On one hand, the configuration file already has a default setting. On the other hand, PDF file upload is now supported.

rsrini7 commented:
Hi Barun Saha,

1. Removed the Mixtral model which is not great in generating slides. Agree
2. localhost reference from OpenRouter optional env. Removed to avoid confusion.
3. Input field introduced for max input text. The default was just 400. but where as i have to give more than 400 and have to manually change in code and restart.

barun-saha commented:
There is no problem in setting `GlobalConfig.LLM_MODEL_MAX_INPUT_LENGTH` to a higher default, say 1000 characters. However, what I'm struggling to understand is why would/should a user choose a value lower than the max value allowed, as seen in the code below:

```python
# Only set session state if not already set by user
if 'llm_max_input_length' not in st.session_state:
    st.session_state['llm_max_input_length'] = 400
llm_max_input_length = st.sidebar.number_input(
    'Max LLM Input Length (characters)',
    min_value=100,
    max_value=20000,
    step=100,
    key='llm_max_input_length',
    help='Maximum number of characters allowed for the LLM input prompt.'
)
```

For argument's sake, `GlobalConfig.LLM_MODEL_MAX_INPUT_LENGTH` can be just set to 20K.

Also, 20K characters seem very high. The text inputs are generally way much shorter. Is there any use case for 20K characters in the input?

rsrini7 commented:
Ok Got It. updated the code. Please check now.

rsrini7 commented:
i had usecase of 10K which again generated by LLM for my project readme with detailed doc and steps. thats why made bit longer input. if its not common usecase. we can keep it 1000 if you agree.

barun-saha commented:
Thanks. I'll take a look.

BTW, SlideDeck AI also supports up to 50 pages of PDF file. You can give that a try for your use case.

barun-saha commented on global_config.py:
```diff
@@ -72,21 +74,30 @@ class GlobalConfig:
+            'description': 'OpenAI GPT-3.5 Turbo (via OpenRouter)',
+            'max_new_tokens': 2048,
+        },
+        '[or]openrouter/gpt-4-omni': {
```
Is this name correct? Or should it be `openai/gpt-4o`? Throws an error saying invalid model ID.

rsrini7 replied:
Thanks for the finding typo. i replaced with gemini model as an example.

barun-saha commented on global_config.py:
```diff
@@ -72,21 +74,30 @@ class GlobalConfig:
             'max_new_tokens': 4096,
             'paid': True,
         },
+        '[or]openai/gpt-3.5-turbo': {
```
- Can you please sort the list items alphabetically? OpenRouter model names should appear before Together AI.
- In addition, can you please update the table of LLMs in the `README` file? There are some references in the texts in the app, too.

barun-saha commented on global_config.py:
```diff
@@ -72,21 +74,30 @@ class GlobalConfig:
         '- **[hf]**: Hugging Face Inference API\n'
-        '- **[to]**: Together AI\n\n'
+        '- **[to]**: Together AI\n'
+        '- **[or]**: OpenRouter\n\n'
```
Please move OpenRouter before Together AI here as well.

rsrini7 replied:
done

barun-saha commented on helpers/llm_helper.py:
```diff
@@ -11,7 +11,8 @@
 from urllib3.util import Retry
 from langchain_core.language_models import BaseLLM, BaseChatModel
-
+from langchain_openai import ChatOpenAI
```
Can you please move this import under the `if provider == GlobalConfig.PROVIDER_OPENROUTER:` block to be consistent with other usage in this module?

barun-saha commented on helpers/llm_helper.py:
```diff
@@ -188,6 +189,21 @@ def get_langchain_llm(
+    if provider == GlobalConfig.PROVIDER_OPENROUTER:
+        # Use langchain-openai's ChatOpenAI for OpenRouter
+        logger.debug('Getting LLM via OpenRouter: %s', model)
+        openrouter_api_key = api_key or os.environ.get("OPENROUTER_API_KEY")
```
The `get_langchain_llm()` is invoked with an API key from `app.py`.  So, the `or os.environ.get("OPENROUTER_API_KEY")` portion is not required. So, you can pass `api_key` directly to the `ChatOpenAI()` call without introducing `openrouter_api_key`.

rsrini7 replied:
done

barun-saha commented on helpers/llm_helper.py:
```diff
@@ -188,6 +189,21 @@ def get_langchain_llm(
+        # Use langchain-openai's ChatOpenAI for OpenRouter
+        logger.debug('Getting LLM via OpenRouter: %s', model)
+        openrouter_api_key = api_key or os.environ.get("OPENROUTER_API_KEY")
+        base_url = "https://openrouter.ai/api/v1"
```
- For consistency, can we please put strings inside single quotes?
- Also, should `base_url` be made a module-level constant? Not sure if the API version will change in the future or how to handle it. Probably can address when it happens.

barun-saha commented on app.py:
```diff
@@ -32,6 +33,18 @@
 
 
+# --- API Key Environment Variable Mapping ---
+PROVIDER_ENV_KEYS = {
```
- Configurations are generally maintained in the `global_config.py` module. Can we please move `PROVIDER_ENV_KEYS` there?

rsrini7 replied:
Yes its moved

barun-saha reviewed (changes requested):
The features look good, thanks.

I have suggested a few code changes. Please have a look.

barun-saha commented on app.py:
```diff
@@ -182,14 +195,24 @@ def reset_api_key():
 
-        # The API key/access token
+        # --- Automatically fetch API key from .env if available ---
+        provider_match = re.match(r'\[(.*?)\]', llm_provider_to_use)
```
Also, can you please replace `re.match()` with a compiled regex pattern?
